IPIP-373: Double Hash DHT Spec (#373)
ipfs/specs: IPIP/0373-double-hash-dht.md

5. [Migration](#migration)
6. [Threat Model](#threat-model)
7. [Alternatives](#alternatives-for-dht-reader-privacy)
8. [Open Questions](#open-questions)

## Motivation

IPFS is currently lacking of many privacy protections. One of its principal weaknesses currently lies in the lack of privacy protections for the DHT content routing subsystem. Currently in the IPFS DHT, neither readers (clients retrieving content) nor writers (hosts storing and distributing content) have much privacy with regard to content they consume or publish. It is trivial for a DHT server node to associate the requestor's identity with the accessed content during the routing process. A curious DHT server node, can request the same CIDs to find out what content other users are consuming. Improving privacy in the IPFS DHT has been a strong request from the community for some time.

The changes described in this document introduce a DHT privacy upgrade boosting the reader’s privacy. It will prevent DHT tracking as described above, and add Provider Records Authentication. The proposed modifications also add a slight Writer Privacy improvement as a side effect.

## Detailed Design

**Magic Values**

> - bytes("CR_DOUBLEHASH")
> - bytes("CR_SERVERKEY")

Reviewer: You should pad (zero or nothing to hide numbers) magic values, this allows implementations to cache the magic vector instead of having to handle with block padding and remainder.
```suggestion
- bytes("CR_DOUBLEHASH\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")
- bytes("CR_SERVERKEY\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")
```

> - AES-GCM [varint](https://github.com/multiformats/multicodec/pull/314): `aes-gcm-256 = 0x8040`

Reviewer 2: Is the multicodec correct?

```suggestion
- AES-GCM [varint](https://github.com/multiformats/multicodec/pull/314): `aes-gcm-256 = 0x2000`
```

Author: Yes this format is correct, multicodec is `0x2000` but the varint is `0x8040`. We use the varint on the wire.

aes-gcm-256 PR: https://github.com/multiformats/multicodec/pull/314

The logic is described here: https://github.com/multiformats/unsigned-varint
8192 (0x2000)   => 10000000 01000000 (0x8040)

- Double SHA256 [varint](https://github.com/multiformats/multicodec/blob/master/table.csv#L41): `dbl-sha2-256 = 0x5601`
- A DHT Server returns all of the Provider Records matching to at most **`MatchLimit = 64`** distinct `HASH2`. Magic number explanation in [k-anonymity](#k-anonymity).
- Provider Record Timestamp (`TS`) validity period: `48h`

### Definitions

- **`CID`** is the IPFS [Content IDentifier](https://github.com/multiformats/cid)
- **`MH`** is the [Multihash](https://github.com/multiformats/multihash) contained in a `CID`. It corresponds to the digest of a hash function over some content. `MH` is represented as a 32-byte array.

> - **`HASH2`** is defined as `SHA256(bytes("CR_DOUBLEHASH") || MH)`. It represents the location of the Kademlia keyspace for the Provider Record associated with `CID`. `HASH2` is represented as a 32-byte array. `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)`.

Reviewer: ```suggestion
- **`HASH2`** is defined as `SHA256(SALT_DOUBLEHASH || MH)`. It represents the location of the Kademlia keyspace for the Provider Record associated with `CID`. `HASH2` is represented as a 32-byte array. `HASH2 = SHA256(SALT_DOUBLEHASH || MH)`.
```

- **`multiaddrs`** are the [network addresses](https://github.com/libp2p/specs/tree/master/addressing) associated with a `PeerID`. It represents the location(s) of the peer.
- **`KeyPrefix`** is defined as a prefix of length `l` bits of `HASH2`. `KeyPrefix` is represented by a `byte` concatenated with a variable sized array of bytes, containing at most 32 bytes. The leading `byte` represents the binary representation of `l - 1`, making prefixes of length `256` possible, but not prefixes of length `0`. The trailing byte array is of length `ceil(l/8)` bytes, and its content is the bits prefix right padded with zeros.
- **`ServerKey`** is defined as `SHA256(bytes("CR_SERVERKEY") || MH)`. It is derived from the `MH`. The Content Provider communicates `ServerKey` to the DHT Servers during the Publish Process. The DHT Servers use `ServerKey` to encrypt `TS`, `Signature` and Content Providers `multiaddrs` sent to the Client when some Provider Records match the requested `Prefix`. `ServerKey` is represented as a 32-byte array.

> - **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.

Author: Initially it was to spare bytes on the wire. However, as we need a 12-byte IV for AES-GCM, we could define the IV to be 8 bytes Timestamp (publish time) + 4 random bytes. So we don't use extra capacity on the wire. WDYT @Jorropo ?

> - **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.

Author: 1. IV should be unique for each encryption key (so unless you want to create billions of distinct provider records for the same content within the same second you should be safe)
2. if we use 64 bits timestamps, it shouldn't be a problem to have timestamp in ns

> - **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.

Author: Let's use a custom timestamp representing the number of minutes since 1970-01-01T00:00:00 on 4 bytes. The remaining 8 bytes of the IV are generated randomly.

> - **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.

Reviewer: > Let's use a custom timestamp representing the number of minutes since 1970-01-01T00:00:00 on 4 bytes. The remaining 8 bytes of the IV are generated randomly.

For context outcome of our private conversation:
- 32 bits seconds is not future proof enough.
- 64 bits seconds is overkill
- 64 bits nanoseconds is anoying on programing languages does have only have seconds precision included by default
- 64 bits is also anoying because that only leaves 4 bytes randomly generated in the IV, which makes collisions more likely than I like.
- 38 bits seconds + 7 random is heresy (not power of two words ?)

So here we go with 32 bits minutes precision.

- **`CPPeerID`** is the `PeerID` of the Content Provider for a specific `CID`.

> - **`EncPeerID`** is the result of the encryption of `CPPeerID` using `MH` as encryption key and a random nonce `AESGCM(MH, Nonce, CPPeerID)`. `EncPeerID` contains the [varint](https://github.com/multiformats/multicodec/pull/314) of the encryption algorithm used (AES), the bytes array of the encrypted payload, and the `Nonce`. `Nonce` is a randomly generated 12-byte array. The format of `EncPeerID` is [`0x8040`, `Nonce`, `payload_len`, `AESGCM(MH, Nonce, CPPeerID)`].

Reviewer 2: Is there a specific reason for `payload_len` appearing after nonce? 
If not, to be consistent with multicodec-code prefixed bytes elsewhere (e.g. multihash) would it make sense to move `payload_len` immediately after multicodec code?

```suggestion
- **`EncPeerID`** is the result of the encryption of `CPPeerID` using `MH` as encryption key and a random nonce `AESGCM(MH, Nonce, CPPeerID)`. `EncPeerID` contains the [varint](https://github.com/multiformats/multicodec/pull/314) of the encryption algorithm used (AES), the bytes array of the encrypted payload, and the `Nonce`. `Nonce` is a randomly generated 12-byte array. The format of `EncPeerID` is [`0x2000`, `payload_len`, `Nonce`, `AESGCM(MH, Nonce, CPPeerID)`].
```

Author: @masih yes, we could swap the positions `payload_len` and `Nonce`. As the Nonce has a fixed size, we don't include it in the `payload_len`.

Reference: https://github.com/multiformats/multihash

Reviewer 2: No hard feelings here either works. Looking across other implementations len after code seems to be the popular order. That's all. Thanks Gui

5. DHT servers compute `Hash(MH)`.
6. DHT servers find the 20 closest peers to `Hash(MH)` in XOR distance in their Routing Table.
7. DHT servers return the 20 `PeerID`s and `multiaddrs` of these peers to Client.
8. Client sends a DHT lookup request for `CID` to the closest peers in XOR distance to `Hash(MH)` that it received.
9. Go to 4.
10. The DHT servers storing the Provider Record(s) associated with `MH` send them to Client. (Currently, if a Provider Record has been published less than 30 min before being requested, the DHT servers also send the `multiaddresses` of the Content Provider to Client).
11. If the response from the DHT server doesn't include the `multiaddrs` associated with the Content Providers' `PeerID`s, Client performs a DHT `FindPeer` request to find the `multiaddrs` of the returned `PeerID`s.
12. Client sends a Bitswap request for `CID` to the Content Provider (known `PeerID` and `multiaddrs`).
13. Content Provider sends the requested content back to Client.

### Double Hash DHT design

**Publish Process**
1. Content Provider wants to publish some content with identifier `CID`.

> 2. Content Provider computes `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).

Reviewer: ```suggestion
2. Content Provider computes `HASH2 = SHA256(SALT_DOUBLEHASH || MH)` (`MH` is the MultiHash included in the CID).
```

3. Content Provider starts a DHT lookup request for the 20 closest `PeerID`s in XOR distance to `HASH2`.

> 4. Content Provider encrypts its own `PeerID` (`CPPeerID`) with `MH`, using AES-GCM. `EncPeerID = [0x8040, Nonce, payload_len, AESGCM(MH, Nonce, CPPeerID)]`

Reviewer 2: ```suggestion
4. Content Provider encrypts its own `PeerID` (`CPPeerID`) with `MH`, using AES-GCM. `EncPeerID = [0x2000, payload_len, Nonce, AESGCM(MH, Nonce, CPPeerID)]`
```

5. Content Provider takes the current timestamp `TS`.
6. Content Provider signs `EncPeerID` and `TS` using its private key. `Signature = Sign(privkey, EncPeerID || TS)`

> 7. Content Provider computes `ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)`.

Reviewer: ```suggestion
7. Content Provider computes `ServerKey = SHA256(SALT_SERVERKEY || MH)`.
```

8. Once the lookup request has returned the 20 closest peers, Content Provider sends a Publish request to these DHT servers. The Publish request contains [`HASH2`, `EncPeerID`, `TS`, `Signature`, `ServerKey`].
9. Each DHT server verifies `Signature` against the `PeerID` of the Content Provider used to open the libp2p connection. `Verify(CPPeerID, Signature, EncPeerID || TS)`. It verifies that `TS` is younger than `48h` and isn't in the future. If invalid, send an error to the client. <!-- TODO: define error -->
10. Each DHT server adds an entry in their Provider Store for `HASH2` -> `ServerKey` -> `CPPeerID` -> [`EncPeerID`, `TS`, `Signature`], with `CPPeerID` being the `PeerID` of the Content Provider (see [provider store](#provider-store)). If there is already an entry including `CPPeerID` for `HASH2` -> `ServerKey`, and if the `TS` of the new valid entry is newer than the existing `TS`, overwrite the entry in the Provider Store. Else drop the new entry.
11. Each DHT server confirms to Content Provider that the Provider Record has been successfully added.

> 12. The proces is over once Content Provider has received 20 confirmations.

Reviewer 3: ```suggestion
12. The process is over once Content Provider has received 20 confirmations.
```

> 12. The proces is over once Content Provider has received 20 confirmations.

Reviewer 3: Nit: `Confirmation` sounds to me that the remote peer needs to acknowledge the receipt of the provider record. So this spec requires 20 successful RPC's. However, in the current implementation a significant fraction of the `ADD_PROVIDER` RPC's error out and we basically just live with it.

The previous steps set the stage where all DHT servers are well functioning so this step 12 may be fine here. However, what do you think of a brief discussion about how to handle errors in this final step?

Author: The number of Provide ACK isn't modified by this Spec, it is expected to stay the same as it currently is. There isn't a proper Spec for the DHT yet, so I couldn't add a reference. Do you know what is the current state of the implementation (regarding the number of Provide ACKs that are required) ?

 The only difference in the Provide Operation is the content that is transmitted from the Content Provider to the DHT Servers.

Reviewer 3: There is no required number of ACKs :/ In the current implementation we try to find the 20 closest peers and as soon as we have them we fire 20 concurrent `ADD_PROVIDER` RPCs. If any call fails we just [log the error](https://github.com/libp2p/go-libp2p-kad-dht/blob/76b0f760072c249de070ee285c434064f50bdcc1/routing.go#L430-L434
) and call it day 🙈

Author: This sounds like something to modify in the Provide process directly 👍🏻

```mermaid
sequenceDiagram
    participant CP as Content Provider
    participant DHT
    participant Server as DHT Server

>     Note left of CP: HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)

Reviewer: ```suggestion
    Note left of CP: HASH2 = SHA256(SALT_DOUBLEHASH || MH)
```

    CP->>DHT: FIND_PEERS(HASH2)
    DHT->>CP: [PeerID0, PeerID1, ... PeerID19]

>     Note left of CP: EncPeerID = 0x8040 || Nonce || payload_len || AESGCM(MH, Nonce, CPPeerID)

Reviewer 2: ```suggestion
    Note left of CP: EncPeerID = 0x2000 || payload_len || Nonce || AESGCM(MH, Nonce, CPPeerID)
```

    Note left of CP: Signature = Sign(privkey, EncPeerID || TS)

>     Note left of CP: ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)

Reviewer: ```suggestion
    Note left of CP: ServerKey = SHA256(SALT_SERVERKEY || MH)
```

    par Content Provider to the 20 closest DHT Servers to HASH2
        CP->>Server: HASH2 || EncPeerID || TS || Signature || ServerKey
    
    Note right of Server: Verify(pubkey, Signature, EncPeerID) &&<br>TS - time.now() < 48h
    Note right of Server: On success, add to Provider Store:<br>HASH2 -> ServerKey -> CPPeerID -> [EncPeerID, TS, Signature]

    Server->>CP: Success / Error
    end

    Note left of CP: Wait for 20 Successes
```

**Lookup Process**

> 1. Client computes `HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)` (`MH` is the MultiHash included in the CID).

Reviewer: ```suggestion
1. Client computes `HASH2 = SHA256(SALT_DOUBLEHASH || MH)` (`MH` is the MultiHash included in the CID).
```

2. Client selects a prefix of `HASH2`, `KeyPrefix = HASH2[:l]` for a defined `l` (see [`l` selection](#prefix-length-selection)).
2. Client finds the closest `PeerID`s to `HASH2` in XOR distance in its Routing Table.
3. Client sends a DHT lookup request for `KeyPrefix` to these DHT servers. The request contains a flag to specify whether Client wants the `multiaddrs` associated with the `CPPeerID` or not. <!-- TODO: add multiaddrs request flag -->
4. The DHT servers find the 20 closest `PeerID`s to `KeyPrefix` in XOR distance (see [algorithm](#closest-keys-to-a-key-prefix)). Add these `PeerID`s and their associated multiaddresses (if applicable) to the `message` that will be returned to Client.
5. The DHT servers search if there are entries matching `KeyPrefix` in their Provider Store.

> 6. For all entries `HASH2` of the Provider Store where `HASH2[:len(KeyPrefix)]==KeyPrefix`, add to `message` the following encrypted payload: `EncPeerID || 0x8040 || SERVERNONCE || payload_len || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`, `SERVERNONCE` being a randomly generated 12-byte array, for `multiaddrs` being the multiaddresses associated with `CPPeerID` (if applicable) if the `multiaddrs` were requested by Client. The `multiaddrs` are taken from the DHT Server's lib2p2 peerstore, and may be stale. If more than `MatchLimit` distinct `HASH2`s match the requested `KeyPrefix`, the DHT Server doesn't return any Provider Record, and adds the number of `HASH2` matching `KeyPrefix` along with its own `MatchLimit` to `message`.

Reviewer 2: Questions:
1. We know that `TS` is 32-bits. In a case where addresses are present, how would the client infer where the `Signature` ends and `multiaddrs` start?
2. Pointed out in other comment: would it make sense to use multicodec code prefix for the inner encrypted content? That should then clarify how to parse the unencryted values. 

```suggestion
6. For all entries `HASH2` of the Provider Store where `HASH2[:len(KeyPrefix)]==KeyPrefix`, add to `message` the following encrypted payload: `EncPeerID || 0x2000 || payload_len || SERVERNONCE || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`, `SERVERNONCE` being a randomly generated 12-byte array, for `multiaddrs` being the multiaddresses associated with `CPPeerID` (if applicable) if the `multiaddrs` were requested by Client. The `multiaddrs` are taken from the DHT Server's lib2p2 peerstore, and may be stale. If more than `MatchLimit` distinct `HASH2`s match the requested `KeyPrefix`, the DHT Server doesn't return any Provider Record, and adds the number of `HASH2` matching `KeyPrefix` along with its own `MatchLimit` to `message`.
```

Author: 1. The Signature length depends on the [PeerID key type](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types). Each key type has a fixed length signature output. 
2. What do you refer to as the "inner encrypted content"?

Reviewer 2: 1. great thank you. Maybe a quick sentence on this either here or in the terminology section would be useful.
2. `TS || Signature || multiaddrs`. Prefixed with the codc to specify what it is, it then becomes something like: `0x12345 || TS || Signature || multiaddrs` where `0x12345` is the multicodec of DHT routing record.

Author: For 2. how about adding this multicodec as below: 

`EncPeerID || MULTICODEC || 0x8040 || SERVERNONCE || payload_len || AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)`

This allows to indicate the format of all meta data, unencrypted and encrypted.

7. The DHT servers send `message` to Client.

> 8. Client computes `ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)`.

Reviewer: ```suggestion
8. Client computes `ServerKey = SHA256(SALT_SERVERKEY || MH)`.
```

9. Client tries to decrypt all returned encrypted payloads using `MH` for `EncPeerID` and `ServerKey` for `Enc(ServerKey, TS || Signature || multiaddrs)`. If at least one encrypted payload can be decrypted, go to 12.
10. If the DHT Server's `MatchLimit` and number of matching `HASH2`s was included in the `message`, Client makes multiple DHT lookup requests for longer prefixes (e.g `KeyPrefix||0` and `KeyPrefix||1`). Else Client sends a DHT lookup request for `KeyPrefix` to the closest peers in XOR distance to `HASH2` that it received from the DHT servers.
11. Go to 4.
12. For each decrypted payload, Client decrypts `CPPeerID = Dec(MH, EncPeerID)`.
13. Client verifies that `Signature` verifies with `CPPeerID`: `Verify(CPPeerID, Signature, EncPeerID || TS)`.
14. Client checks that `TS` is younger than `48h`.
15. If none of the decrypted payloads is valid, go to 4.
16. If the decrypted payload doesn't include the `multiaddrs` associated with `CPPeerID`, Client performs a DHT `FindPeer` request to find the `multiaddrs` associated with `CPPeerID`.

> 17. Client sends a Bitswap request for `CID` to the Content Provider (known `CPPeerID` and `multiaddrs`).

Reviewer: This breaks the layers, DHT does not knows about bitswap, all the DHT is tasked to do is map `mh -> {many provider records}`, the DHT spec can't talk about bitswap (other protocols use the DHT such as pubsub, new datatransfer protocols, ...).

Author: The DHT doesn't directly interact with Bitswap. I added this line for more clarity of content retrieval, and showing the overall privacy that is added (as DHT and Bitswap don't use the same identifiers). I'll remove this line

Reviewer: I'm fine with this line if it's more generic like:
> 17. Client contact the Content Provider (known `CPPeerID` and `multiaddrs`) and can exchange data (the DHT may be consumed by various different protocols).

18. Content Provider sends the requested content back to Client.

```mermaid
sequenceDiagram
    participant Client
    participant Server as DHT Server
    participant CP as Content Provider

>     Note left of Client: HASH2 = SHA256(bytes("CR_DOUBLEHASH") || MH)
>     Note left of Client: ServerKey = SHA256(bytes("CR_SERVERKEY") || MH)

Reviewer: ```suggestion
    Note left of Client: HASH2 = SHA256(SALT_DOUBLEHASH || MH)
    Note left of Client: ServerKey = SHA256(SALT_SERVERKEY || MH)
```

    loop in parallel until valid Provider Record found
        Note left of Client: KeyPrefix = HASH2[:l]
        Client->>Server: FIND_CONTENT(KeyPrefix)<br>Optional flags: multiaddrs, metadata
        Note right of Server: message = []
        loop for each of the 20 closest PeerIDs to KeyPrefix in the Routing Table
            Note right of Server: message += PeerID
        end
        loop for each entry matching KeyPrefix in the Provider Store

>             Note right of Server: EncMetadata = 0x8040 || SERVERNONCE || payload_len ||<br> AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)

Reviewer 2: ```suggestion
            Note right of Server: EncMetadata = 0x2000 || payload_len || SERVERNONCE ||<br> AESGCM(ServerKey, SERVERNONCE, TS || Signature || multiaddrs)
```

Enumerating the number of Provider Records in the DHT becomes trivial thank to the Double Hashing and Prefix Requests. Knowledge of the preimage of the requested key isn't required anymore for a valid Kademlia request. Monitoring the number of Provider Records stored in the DHT is a good metric to evaluate the health of the DHT.

### Better Kademlia Routing Table Refresh

As knowledge of the preimage of the requested key isn't necessary in the Double Hashing DHT, nodes gain the ability to request _truly_ random keys in the DHT.

Requesting random keys is necessary for the Kademlia Bucket Refresh Process. On refresh, if a bucket has empty slots, the node will make a request for a random forged key falling in this specific bucket. In the current implementation, as the prefix of a requested key is necessary, Kademlia uses a [list of precomputed preimages](https://github.com/libp2p/go-libp2p-kbucket/blob/master/bucket_prefixmap.go), 1 matching every 15-bits key prefix. Hence, the random forged key, is never random, its definition set is the list of precomputed preimages, and not the full keyspace. This can lead to degraded performance and security vulnerabilities.

> Double Hashing enables the nodes to select a _truly_ random key from the Kademlia keyspace (limited by the randomness algorithm) matching the appropriate bucket.The 456KB [list of precomputed preimages](https://github.com/libp2p/go-libp2p-kbucket/blob/master/bucket_prefixmap.go) can be removed from the IPFS source code, once the migration to the Double Hashing DHT is complete.

Reviewer: ```suggestion
Double Hashing enables the nodes to select a _truly_ random key from the Kademlia keyspace (limited by the randomness algorithm) matching the appropriate bucket.The 456KB [list of precomputed preimages](https://github.com/libp2p/go-libp2p-kbucket/blob/2e310782ef7bc42d9af3e948fcf21d92b97e56ea/bucket_prefixmap.go) can be removed from the IPFS source code, once the migration to the Double Hashing DHT is complete.
```
if we are gonna remove this, this link is gonna break in the future.

Ephemeral `PeerID`s references:
- https://github.com/libp2p/libp2p/issues/37

The other alternative to increase the Reader Privacy level in the IPFS DHT is the use of Mixnets such as Tor or I2P. Mixnets usually provide an excellent Reader- and Writer Privacy level, but lookup latency is significantly higher. Hence the use of Mixnets is generally not good for all use cases, but only when strong privacy guarantees are required. Mixnets can easily be built on top of the Double Hash DHT to maximize user Privacy. IPFS users willing to remain pseudonymous could use the existing Tor network to hide their identity. Another alternative could be to create a Mixnet out of the IPFS network, e.g include mixing capabilities in every libp2p host. There has been some ongoing work on IPFS-Tor integration.

Mixnets references:

> - Berty's [go-libp2p-tor-transport](https://github.com/berty/go-libp2p-tor-transport)

Reviewer: This is not a really a viable solution, in theory it work but this is as the cost of high latency.
This also relies on politics (/ the tor consensus algorithm) and not encryption to secure the requests.

Author: I agree, I just wanted to listed other alternatives, even though they are less viable

- It may be fine to use `TS` as Nonce/IV for the Provider Record encryption (`EncPeerID = AESGCM(MH, Nonce, CPPeerID)`), it spares bytes on the wire. If `TS` is the number of milli- or nano-seconds that have passed since `1970-01-01T00:00:00`, this number easily fits in the 12 bytes IV. Moreover it is very unlikely that 2 nodes perform an encryption using the same key (for the same content) at the exact same milli- or nano-second. Using TS as nonce would spare 4 bytes (`TS` size) on the wire when publishing content to the DHT, and 4 bytes for each Provider Record matching `Prefix` for all requests. However the information about when the Provider Record was published (already known to the DHT Servers storing the Provider Record) would be publicly available. Anyone enumerating DHT Provider Records would be able to read it.
- As multiple `HASH2` match each `Prefix` and the Client is only interested in a single one, should we send the `HASH2` along with each encrypted provider record (network load overhead) or let the Client try to decrypt all payloads and see for themselves which one opens (cpu overhead)?

> 

Reviewer: - XOF ?
- What is your upgrade story once the hashing function you picked is found vulnerable ?
  - None impacts are not that bad ?
  - Multihash upgrade ?
  - Breaking protocol upgrade ?

Author: XOF = eXtendable Output Funtion? To what extend would this be useful?

Author: If the Hash Function is found vulnerable to 1st preimage attack only, all content providers must republish all content at a different keyspace location (a different HASH2) determined by the new hash function. During the migration period, client request both old HASH2 and new HASH2 to make sure to find the provider record.

If the Hash Function used to derive `MH` from the content is found vulnerable to 1st or 2nd preimage attack, all identifiers have to be migrated, and the same process as above should be followed.

Does this address your questions @Jorropo ?

Reviewer: > XOF = eXtendable Output Funtion? To what extend would this be useful?

IDK that was a question, since you are truncating hashes maybe that would help ?
(probably not) ?

> If the Hash Function is found vulnerable to 1st preimage attack only, all content providers must republish all content at a different keyspace location (a different HASH2) determined by the new hash function. During the migration period, client request both old HASH2 and new HASH2 to make sure to find the provider record.

If the Hash Function used to derive MH from the content is found vulnerable to 1st or 2nd preimage attack, all identifiers have to be migrated, and the same process as above should be followed.

SGTM

- **`EncMetadata`** contains all data associated with `EncMetadataPayload`. `EncMetadata` is composed of the length of `EncMetadataPayload` <!-- TODO: determine len format-->, the `ServerNonce` and `EncMetadataPayload`, and has the following format `EncMetadata = payload_len || ServerNonce || EncMetadataPayload`. This format is defined by `dht-pr-format-v0`.
- **`PRShortIdentifier`** is a uniquely and minimally identifying multiple `HASH2`s matching a `KeyPrefix`. Its format is defined [below](#wire-formats).
- **Provider Record** is defined as a pointer to the storage location of some content identified by `CID` or `HASH2`. A Provider Record consists on the following fields: [`EncPeerID`, `TS`, `Signature`].
- **Provider Store** is the data structure on the DHT Servers used to store the Provider Records. Its structure is a [nested key-value store](#provider-store): `HASH2` -> `ServerKey` -> `CPPeerID` -> Provider Record.

> ### Wire formats

Reviewer 4: Maybe you're waiting for the implementation to be further along, but having the proposed wire formats for the RPC calls (and effectively a new version of https://github.com/libp2p/specs/tree/master/kad-dht for this new version) would be helpful to understand some of the concrete changes and give feedback.

If you're going to be doing a network fork here it might also be easy enough to make some other wire format changes that will improve how the network operates.

**Lookup Process**
1. Client computes `HASH2 = SHA256(SALT_DOUBLEHASH || MH)` (`MH` is the MultiHash included in the CID).
2. Client selects a prefix of `HASH2`, `KeyPrefix = HASH2[:l]` for a defined `l` (see [`l` selection](#prefix-length-selection)).

> <<<<<<< HEAD

Reviewer: :eye: :eye:
